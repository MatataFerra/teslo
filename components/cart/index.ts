export * from "./CartList";
export * from "./OrderSummary";
export * from "./CartProduct";
export * from "./CartSkeleton";
