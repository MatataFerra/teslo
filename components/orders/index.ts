export * from "./PaymentButtons";
export * from "./StatusOrderIcon";
