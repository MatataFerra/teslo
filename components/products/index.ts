export * from "./ProductList";
export * from "./ProductCard";
export * from "./ProductSlideShow";
export * from "./SizeSelector";
export * from "./CtaButtons";
