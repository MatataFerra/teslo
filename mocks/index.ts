export * from "./handlres";
export * from "./productMock";
export * from "./server";
