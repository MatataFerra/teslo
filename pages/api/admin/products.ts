import type { NextApiRequest, NextApiResponse } from "next";
import { db } from "../../../database";
import { IProduct, IProductSize } from "../../../interfaces";
import { Product, ProductSize } from "../../../models";
import { isValidObjectId } from "mongoose";
import { v2 as cloudinary } from "cloudinary";

type Data = { message: string } | IProductSize[] | IProductSize;
cloudinary.config(process.env.CLOUDINARY_URL || "");

export default function handler(req: NextApiRequest, res: NextApiResponse<Data>) {
  if (req.method === "GET") {
    return getProducts(req, res);
  }

  if (req.method === "POST") {
    return createProduct(req, res);
  }

  if (req.method === "PUT") {
    return updateProduct(req, res);
  }

  res.status(400).json({ message: "Method not allowed" });
}

const getProducts = async (req: NextApiRequest, res: NextApiResponse<Data>) => {
  await db.connect();
  const products = await ProductSize.find({}).sort({ title: "asc" }).lean();
  await db.disconnect();

  // actualizar imagenes

  const updatedProducts = products.map((product) => {
    product.images = product.images.map((image) => {
      const regex = new RegExp(/https?/g);
      return regex.test(image) ? image : `${process.env.HOST_NAME}products/${image}`;
    });

    return product;
  });

  res.status(200).json(updatedProducts);
};

const updateProduct = async (req: NextApiRequest, res: NextApiResponse<Data>) => {
  const { _id = "", images = [] } = req.body as IProductSize;

  if (!isValidObjectId(_id)) {
    return res.status(400).json({ message: "Invalid id" });
  }

  if (images.length < 2) {
    return res.status(400).json({ message: "You must upload at least 2 images" });
  }

  // TODO: hay que evitar que se cree la ruta de la imagen

  try {
    await db.connect();
    const product = await ProductSize.findById(_id);

    if (!product) {
      await db.disconnect();
      throw new Error("Product not found");
    }

    // TODO: eliminar fotos cloudinary
    product.images.forEach(async (image) => {
      if (!images.includes(image)) {
        const [fileId] = image.substring(image.lastIndexOf("/") + 1).split(".");
        await cloudinary.uploader.destroy(fileId);
      }
    });

    await product.updateOne(req.body);
    await db.disconnect();

    return res.status(200).json(product);
  } catch (error) {
    console.log(error);
    await db.disconnect();
    return res.status(400).json({ message: "Error updating product" });
  }
};

const createProduct = async (req: NextApiRequest, res: NextApiResponse<Data>) => {
  const { images = [] } = req.body as IProductSize;

  if (images.length < 2) {
    return res.status(400).json({ message: "You must upload at least 2 images" });
  }

  // TODO: hay que evitar que se cree la ruta de la imagen

  try {
    await db.connect();
    const existingProduct = await ProductSize.findOne({ slug: req.body.slug });

    if (existingProduct) {
      await db.disconnect();
      throw new Error("Product already exists with this slug");
    }

    const product = new ProductSize(req.body);
    product.inStock = product.sizes.reduce((acc, size) => acc + size.stock, 0);
    await product.save();
    await db.disconnect();

    return res.status(201).json(product);
  } catch (error) {
    console.log(error);
    await db.disconnect();
    return res.status(400).json({ message: "Error creating product" });
  }
};
