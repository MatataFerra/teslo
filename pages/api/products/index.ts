import type { NextApiRequest, NextApiResponse } from "next";
import { db, SHOP_CONSTANTS } from "../../../database";
import { IProduct, IProductSize } from "../../../interfaces";
import { Product, ProductSize } from "../../../models";

type Data = { message: string } | IProductSize[];

export default function handler(req: NextApiRequest, res: NextApiResponse<Data>) {
  if (req.method === "GET") {
    return getProducts(req, res);
  }

  res.status(400).json({ message: "Method not allowed" });
}

const getProducts = async (req: NextApiRequest, res: NextApiResponse<Data>) => {
  const { gender = "all" } = req.query;
  let condition = {};

  if (gender !== "all" && SHOP_CONSTANTS.validGenders.includes(gender.toString())) {
    condition = { gender };
  }

  await db.connect();

  const products = await ProductSize.find(condition).select("title images price inStock slug -_id").lean();
  await db.disconnect();

  const updatedProducts = products.map((product) => {
    product.images = product.images.map((image) => {
      return image.includes("http") ? image : `${process.env.HOST_NAME}products/${image}`;
    });

    return product;
  });

  return res.status(200).json(updatedProducts);
};
