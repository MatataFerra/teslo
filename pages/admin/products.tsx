import { NextPage } from "next";
import NextLink from "next/link";

import { AdminLayout } from "../../components/layouts";
import { AddOutlined, CategoryOutlined } from "@mui/icons-material";
import { Grid, CardMedia, Link, Box, Button, Tooltip } from "@mui/material";
import { DataGrid, GridColDef, GridValueGetterParams } from "@mui/x-data-grid";
import { IProductSize, ISizeStock, ISize } from "../../interfaces";
import useSWR from "swr";
import { ErrorComponent, FullScreenLoading } from "../../components/ui";

enum ValidSize {
  XS = "XS",
  S = "S",
  M = "M",
  L = "L",
  XL = "XL",
  XXL = "XXL",
  XXXL = "XXXL",
}

const parseSize = (sizes: ISizeStock[], validSize: ISize = "S") => {
  return sizes
    .filter((m: any) => m.size === validSize)
    .map((size: any) => {
      const stock = `${size.stock}`;
      return stock;
    });
};

const columns: GridColDef[] = [
  {
    field: "img",
    headerName: "Foto",
    renderCell: ({ row }: GridValueGetterParams) => {
      return (
        <a href={`/product/${row.slug}`} target='_blank' rel='noopener noreferrer'>
          <Tooltip title='Ver producto' arrow placement='right'>
            <CardMedia component='img' className='fadeIn' image={row.img} />
          </Tooltip>
        </a>
      );
    },
  },
  {
    field: "title",
    headerName: "Título del producto",
    width: 250,
    renderCell: ({ row }: GridValueGetterParams) => {
      return (
        <NextLink href={`/admin/products/${row.slug}`} passHref>
          <Tooltip title='Editar producto' arrow placement='right'>
            <Link underline='always'>{row.title}</Link>
          </Tooltip>
        </NextLink>
      );
    },
  },
  { field: "gender", headerName: "Género" },
  { field: "type", headerName: "Tipo" },
  { field: "inStock", headerName: "Inventario" },
  { field: "price", headerName: "Precio" },
  {
    field: "xs",
    headerName: ValidSize.XS,
    width: 75,
    renderCell: ({ row }: GridValueGetterParams) => {
      return !row.xs.length ? "-" : row.xs;
    },
  },

  {
    field: "s",
    headerName: ValidSize.S,
    width: 75,
    renderCell: ({ row }: GridValueGetterParams) => {
      return !row.s.length ? "-" : row.s;
    },
  },

  {
    field: "m",
    headerName: ValidSize.M,
    width: 75,
    renderCell: ({ row }: GridValueGetterParams) => {
      return !row.m.length ? "-" : row.m;
    },
  },

  {
    field: "l",
    headerName: ValidSize.L,
    width: 75,
    renderCell: ({ row }: GridValueGetterParams) => {
      return !row.l.length ? "-" : row.l;
    },
  },

  {
    field: "xl",
    headerName: ValidSize.XL,
    width: 75,
    renderCell: ({ row }: GridValueGetterParams) => {
      return !row.xl.length ? "-" : row.xl;
    },
  },

  {
    field: "xxl",
    headerName: ValidSize.XXL,
    width: 75,
    renderCell: ({ row }: GridValueGetterParams) => {
      return !row.xxl.length ? "-" : row.xxl;
    },
  },

  {
    field: "xxxl",
    headerName: ValidSize.XXXL,
    width: 75,
    renderCell: ({ row }: GridValueGetterParams) => {
      return !row.xxxl.length ? "-" : row.xxxl;
    },
  },
];

const ProductsPage: NextPage = () => {
  const { data, error } = useSWR<IProductSize[]>("/api/admin/products");

  if (!data && !error) return <FullScreenLoading />;
  if (error) return <ErrorComponent />;

  const rows = data!.map((product) => {
    return {
      id: product._id,
      img: product.images[0],
      title: product.title,
      gender: product.gender,
      type: product.type,
      inStock: product.sizes.reduce((acc, curr) => acc + curr.stock, 0),
      price: product.price,
      xs: parseSize(product.sizes, ValidSize.XS),
      s: parseSize(product.sizes, ValidSize.S),
      m: parseSize(product.sizes, ValidSize.M),
      l: parseSize(product.sizes, ValidSize.L),
      xl: parseSize(product.sizes, ValidSize.XL),
      xxl: parseSize(product.sizes, ValidSize.XXL),
      xxxl: parseSize(product.sizes, ValidSize.XXXL),
      slug: product.slug,
    };
  });

  return (
    <AdminLayout
      title={`Productos (${data?.length ?? 0})`}
      subtitle={"Mantenimiento de órdenes"}
      icon={<CategoryOutlined />}
      back>
      <Box display='flex' justifyContent='end' sx={{ mb: 2 }}>
        <Button startIcon={<AddOutlined />} variant='contained' color='primary' href='/admin/products/new'>
          Nuevo producto
        </Button>
      </Box>

      <Grid container className='fadeIn'>
        <Grid item xs={12} sx={{ height: 650, width: "100%" }}>
          <DataGrid rows={rows} columns={columns} pageSize={10} rowsPerPageOptions={[10]} />
        </Grid>
      </Grid>
    </AdminLayout>
  );
};

export default ProductsPage;
