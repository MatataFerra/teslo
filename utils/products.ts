import { IProductSize } from "../interfaces/products";

export const parseImagesOnProducts = (products: IProductSize[]) => {
  /**Parse images. If you have images on cloud and local you must  need to parse the array  */
  return products.map((product) => {
    product.images = product?.images?.map((image) => {
      return image.includes("http") ? image : `${process.env.NEXT_PUBLIC_VERCEL_URL}products/${image}`;
    });

    return product;
  });
};
