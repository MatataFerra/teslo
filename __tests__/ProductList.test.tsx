import { render, fireEvent, waitFor, prettyDOM } from "@testing-library/react";
import { ProductList } from "../components/products";
import { productsMock } from "../mocks";

describe("ProductList", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("should render correctly", () => {
    const { asFragment } = render(<ProductList products={productsMock} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should render correctly with no products", () => {
    const { asFragment } = render(<ProductList products={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("should redirect to slug when click on image", () => {
    const { getByRole } = render(<ProductList products={productsMock} />);
    const product = getByRole("link");

    expect(product.getAttribute("href")).toEqual(`/product/${productsMock[0].slug}`);

    // expect("/product/slug").toBe(`/product/${productsMock[0].slug}`);
  });
});
