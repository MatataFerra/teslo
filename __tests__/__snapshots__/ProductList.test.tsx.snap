// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 css-1xd74lj-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-6xddht-MuiTypography-root-MuiLink-root"
          data-testid="link-slug-product-card"
          href="/product/slug"
          role="link"
        >
          <button
            class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
            tabindex="0"
            type="button"
          >
            <img
              alt="title"
              class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img fadeIn css-o69gx8-MuiCardMedia-root"
              src="https://via.placeholder.com/300"
            />
            <span
              class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
      <div
        class="fadeIn MuiBox-root css-164r41r"
      >
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 70%; height: 20px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
          style="width: 10%; height: 20px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProductList should render correctly with no products 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
  />
</DocumentFragment>
`;

exports[`ProductList should render correctly with no products 2`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-4 css-480o17-MuiGrid-root"
  />
</DocumentFragment>
`;
